<a href="https://colab.research.google.com/github/anabiarochar/Mestrado_AlgoritmosProgramacao/blob/main/Projeto1/3_Inser%C3%A7%C3%A3o.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Projeto 1 - Algoritmo de Ordenação Inserção (Insertion Sort)

Abaixo estão os algorítmos para ordenação, utilizando o método inserção de ordenação, com entradas de vetores de tamanhos 1.000, 10.000 e 100.000
- aleatórias
- ordenadas em forma crescente
- ordenadas em forma decrescente

Abaixo de cada algorítmo estão os tempos observados somando 4 tentativas.

##Vetor em ordem aleatória com 1000 elementos:


In [4]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_aleatorio.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")

# Método Inserção
def insertionSort(A, n):
    for j in range(1, n):
        x = A[j]
        i = j - 1

        while i >= 0 and A[i] > x:
            A[i + 1] = A[i]
            i -= 1
        A[i + 1] = x
    return A

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_para_ordenar = vetor.copy()
    vetor_ordenado = insertionSort(vetor_para_ordenar, len(vetor_para_ordenar))
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.048388 segundos
[1, 3, 8, 9, 9, 10, 10, 10, 10, 12]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.025019 segundos  
Tentativa 2 - Tempo de execução: 0.024606 segundos  
Tentativa 3 - Tempo de execução: 0.021900 segundos  
Tentativa 4 - Tempo de execução: 0.048388 segundos  

##Vetor em ordem aleatória com 10000 elementos:


In [6]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_aleatorio.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")

# Método Inserção
def insertionSort(A, n):
    for j in range(1, n):
        x = A[j]
        i = j - 1

        while i >= 0 and A[i] > x:
            A[i + 1] = A[i]
            i -= 1
        A[i + 1] = x
    return A

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_para_ordenar = vetor.copy()
    vetor_ordenado = insertionSort(vetor_para_ordenar, len(vetor_para_ordenar))
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 2.391673 segundos
[3, 3, 4, 8, 9, 9, 9, 11, 12, 12]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 3.796516 segundos  
Tentativa 2 - Tempo de execução: 2.521852 segundos  
Tentativa 3 - Tempo de execução: 2.684992 segundos  
Tentativa 4 - Tempo de execução: 2.391673 segundos  

##Vetor em ordem aleatória com 100000 elementos:


In [7]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_aleatorio.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")

# Método Inserção
def insertionSort(A, n):
    for j in range(1, n):
        x = A[j]
        i = j - 1

        while i >= 0 and A[i] > x:
            A[i + 1] = A[i]
            i -= 1
        A[i + 1] = x
    return A

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_para_ordenar = vetor.copy()
    vetor_ordenado = insertionSort(vetor_para_ordenar, len(vetor_para_ordenar))
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 436.935766 segundos
[2, 5, 6, 6, 8, 9, 10, 13, 14, 14]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 400.612380 segundos  
Tentativa 2 - Tempo de execução: 455.484334 segundos  
Tentativa 3 - Tempo de execução: 419.123910 segundos
Tentativa 4 - Tempo de execução: 436.935766 segundos

##Vetor em ordem crescente com 1000 elementos:


In [11]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_crescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")

# Método Inserção
def insertionSort(A, n):
    for j in range(1, n):
        x = A[j]
        i = j - 1

        while i >= 0 and A[i] > x:
            A[i + 1] = A[i]
            i -= 1
        A[i + 1] = x
    return A

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_para_ordenar = vetor.copy()
    vetor_ordenado = insertionSort(vetor_para_ordenar, len(vetor_para_ordenar))
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.000129 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.004086 segundos  
Tentativa 2 - Tempo de execução: 0.000210 segundos  
Tentativa 3 - Tempo de execução: 0.000184 segundos  
Tentativa 4 - Tempo de execução: 0.000129 segundos  

##Vetor em ordem crescente com 10000 elementos:


In [14]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_crescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")

# Método Inserção
def insertionSort(A, n):
    for j in range(1, n):
        x = A[j]
        i = j - 1

        while i >= 0 and A[i] > x:
            A[i + 1] = A[i]
            i -= 1
        A[i + 1] = x
    return A

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_para_ordenar = vetor.copy()
    vetor_ordenado = insertionSort(vetor_para_ordenar, len(vetor_para_ordenar))
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.001529 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.001235 segundos  
Tentativa 2 - Tempo de execução: 0.001235 segundos  
Tentativa 3 - Tempo de execução: 0.001312 segundos  
Tentativa 4 - Tempo de execução: 0.001529 segundos  

##Vetor em ordem crescente com 100000 elementos:


In [18]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_crescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")

# Método Inserção
def insertionSort(A, n):
    for j in range(1, n):
        x = A[j]
        i = j - 1

        while i >= 0 and A[i] > x:
            A[i + 1] = A[i]
            i -= 1
        A[i + 1] = x
    return A

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_para_ordenar = vetor.copy()
    vetor_ordenado = insertionSort(vetor_para_ordenar, len(vetor_para_ordenar))
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.025860 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.023700 segundos  
Tentativa 2 - Tempo de execução: 0.013678 segundos  
Tentativa 3 - Tempo de execução: 0.013505 segundos  
Tentativa 4 - Tempo de execução: 0.025860 segundos  

##Vetor em ordem decrescente com 1000 elementos:


In [22]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_decrescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")

# Método Inserção
def insertionSort(A, n):
    for j in range(1, n):
        x = A[j]
        i = j - 1

        while i >= 0 and A[i] > x:
            A[i + 1] = A[i]
            i -= 1
        A[i + 1] = x
    return A

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_para_ordenar = vetor.copy()
    vetor_ordenado = insertionSort(vetor_para_ordenar, len(vetor_para_ordenar))
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.041443 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.042606 segundos  
Tentativa 2 - Tempo de execução: 0.040912 segundos  
Tentativa 3 - Tempo de execução: 0.041046 segundos  
Tentativa 4 - Tempo de execução: 0.041443 segundos  

##Vetor em ordem decrescente com 10000 elementos:


In [26]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_decrescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")

# Método Inserção
def insertionSort(A, n):
    for j in range(1, n):
        x = A[j]
        i = j - 1

        while i >= 0 and A[i] > x:
            A[i + 1] = A[i]
            i -= 1
        A[i + 1] = x
    return A

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_para_ordenar = vetor.copy()
    vetor_ordenado = insertionSort(vetor_para_ordenar, len(vetor_para_ordenar))
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 4.697968 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 5.820077 segundos  
Tentativa 2 - Tempo de execução: 5.878299 segundos  
Tentativa 3 - Tempo de execução: 5.184855 segundos  
Tentativa 4 - Tempo de execução: 4.697968 segundos  

##Vetor em ordem decrescente com 100000 elementos:


In [28]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_decrescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")

# Método Inserção
def insertionSort(A, n):
    for j in range(1, n):
        x = A[j]
        i = j - 1

        while i >= 0 and A[i] > x:
            A[i + 1] = A[i]
            i -= 1
        A[i + 1] = x
    return A

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_para_ordenar = vetor.copy()
    vetor_ordenado = insertionSort(vetor_para_ordenar, len(vetor_para_ordenar))
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 558.093165 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 554.920139 segundos  
Tentativa 2 - Tempo de execução: 550.336984 segundos  
Tentativa 3 - Tempo de execução: 555.754223 segundos  
Tentativa 4 - Tempo de execução: 558.093165 segundos  

##Resumo dos Tempos de Execução

| Caso | n | Tentativa 1 | Tentativa 2 | Tentativa 3 | Tentativa 4 | Média |
|---|---:|---:|---:|---:|---:|---:|
| Aleatório | 1.000 | 0.025019 | 0.024606 | 0.021900 | 0.048388 | 0.029978 |
| Aleatório | 10.000 | 3.796516 | 2.521852 | 2.684992 | 2.391673 | 2.848758 |
| Aleatório | 100.000 | 400.612380 | 455.484334 | 419.123910 | 436.935766 | 428.039098 |
| Crescente | 1.000 | 0.004086 | 0.000210 | 0.000184 | 0.000129 | 0.001152 |
| Crescente | 10.000 | 0.001235 | 0.001235 | 0.001312 | 0.001529 | 0.001328 |
| Crescente | 100.000 | 0.023700 | 0.013678 | 0.013505 | 0.025860 | 0.019186 |
| Decrescente | 1.000 | 0.042606 | 0.040912 | 0.041046 | 0.041443 | 0.041502 |
| Decrescente | 10.000 | 5.820077 | 5.878299 | 5.184855 | 4.697968 | 5.395300 |
| Decrescente | 100.000 | 554.920139 | 550.336984 | 555.754223 | 558.093165 | 554.776128 |
